### PR TITLE
Feat/issue 1

### PR DESCRIPTION
**Pull request by shhdharmen**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngneat/falso/blob/master/CONTRIBUTING.md#commit
~~[ ] Tests for the changes have been added (for bug fixes / features)~~
~~[ ] Docs have been added / updated (for bug fixes / features)~~

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #1 

## What is the new behavior?

Adds locale support through executors and google translate. To generate translations, one will need to follow below steps:

### Modify `workspace.json` file:

Check / change below options under "translate" object:

- `path` - List of file paths of which translation needs to be generated
- `languages` - List of languages to which translations will be generated. For full list refer to: <https://cloud.google.com/translate/docs/languages>
- `output` - Output directory when translation files will be generated

### Run `translate` script

Once you've reviewed the options, you just need to run below script from root of the project:

```bash
npm run translate
```

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: tools/executors/translate/generate.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  const stringLiterals = getStringLiterals(sourceFile);
+  return { stringLiterals, sourceFile };
+}
+
+function getStringLiterals(node: ts.Node) {
+  let str: string[] = [];
+  node.forEachChild((childNode) => {
+    if (ts.isArrayLiteralExpression(childNode)) {
+      childNode.forEachChild((stringNode) => {
+        if (ts.isStringLiteral(stringNode)) {
```
Comment (NetanelBasal) [inline tools/executors/translate/generate.ts:23]: Can you please add examples in a comment that show which code you're trying to find in each line?

Btw, consider using ts-query.

File: tools/executors/translate/write.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+  appendFileSync(index, exportStatement, { encoding: 'utf-8' });
+}
+
+export function deleteLanguageIndexFiles(options: EchoExecutorOptions) {
+  for (const language of options.languages) {
+    console.info(`Deleting index file for language: ${language}`);
+    const outputDir = join(options.output, language);
+    if (!existsSync(outputDir)) {
+      mkdirSync(outputDir);
```
Comment (NetanelBasal) [inline tools/executors/translate/write.ts:84]: The name of the function is `delete`, but you're creating it here.

File: workspace.json [hunk 3]
```diff
@@ -1,6 +1,20 @@
...
-    "falso": "packages/falso"
+    "falso": {
+      "root": "packages/false",
+      "sourceRoot": "packages/false/src",
+      "projectType": "package",
+      "targets": {
+        "translate": {
+          "executor": "./tools/executors/translate:translate",
+          "options": {
+            "path": ["packages/falso/src/lib/abbreviation.ts"],
```
[the remark this answers is not in the document]
Comment (shhdharmen) [reply workspace.json]: Done

File: tools/executors/translate/generate.ts (hunk 1 above)
Comment (shhdharmen) [reply tools/executors/translate/generate.ts]: Done

File: tools/executors/translate/index.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+      const fileStats = statSync(rootFilePath);
+
+      const isFile = fileStats.isFile();
+      if (isFile) {
+        console.info(`Reading content from ${rootFilePath}`);
+
+        const { stringLiterals, sourceFile } =
+          generateStringLiteralsAndSourceFile(rootFilePath);
+
+        if (stringLiterals && stringLiterals.length) {
```
[the remark this answers is not in the document]
Comment (shhdharmen) [reply tools/executors/translate/index.ts]: Done

File: tools/executors/translate/write.ts (hunk 2 above)
Comment (shhdharmen) [reply tools/executors/translate/write.ts:83]: Renamed function name

**Comment by NetanelBasal** [comment]
Sorry for the hassle; I believe it will be better to take the following approach.

Let's take the `abbreviation` function example:

```ts
import { rand } from './core';

export function abbreviation() {
  return rand([...]]);
}
```

- Create a JSON file named `abbreviation.i18n.json` and extract the static data to the file:

```
// abbreviation.i18n.json

{
  data: [....]
}
````

- Change the `abbreviation.ts` to use it:

```ts
import { rand } from './core';
import data from './abbreviation.i18n.json';

export function abbreviation() {
  return rand(data);
}
```

We can seamlessly go over all the files named `i18n` and translate the data without using AST. 

You can start with the `abbreviation` and later create a script that transfers the others to this style.

**Comment by shhdharmen** [comment]
@NetanelBasal made changes as per your last comment.

File: tools/executors/translate/README.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+2. [Generate keys](https://cloud.google.com/translate/docs/setup#creating_service_accounts_and_keys) and download them
+3. Place JSON at `./files/google-cloud-secret.json`.
+
+## Usage
+
+### Modify `workspace.json` file
+
+Check / change below options under "translate" object:
+
+- `path` - List of file paths of which translation needs to be generated
```
Comment (NetanelBasal) [inline tools/executors/translate/README.md:15]: There is no need for a path anymore. We made the i18n addon to know which files to use seamlessly.

Comment (shhdharmen) [reply tools/executors/translate/README.md:15]: Until we create json for each file, it's needed.

File: tools/executors/translate/generate.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
+import { readFileSync } from 'fs';
+import * as ts from 'typescript';
+import { tsquery } from '@phenomnomnominal/tsquery';
+
+export function generateStringLiteralsAndSourceFile(rootFilePath: string) {
```
[the remark this answers is not in the document]
Comment (shhdharmen) [reply tools/executors/translate/generate.ts:5]: Until we create json for each file, it's needed.

[the remark this answers is not in the document]
Comment (NetanelBasal) [reply tools/executors/translate/generate.ts:5]: So I think the JSON task should be merged before this one.

[the remark this answers is not in the document]
Comment (shhdharmen) [reply tools/executors/translate/generate.ts:5]: Running `npm run translate` will do everything:

1. Create i18n json at lib dir
2. Use data from i18n json in lib dir
3. Create ts and i18n json in translation folders

**Pull request closed without merging**